src/ImageSharp/Formats/Jpeg/Components/FastFloatingPointDCT.cs

```diff
@@ -43,259 +42,187 @@ internal static partial class FastFloatingPointDCT
-#if SUPPORTS_RUNTIME_INTRINSICS
-            Debug.Assert(Avx.IsSupported, "AVX is required to execute this method");
-
-            Vector256<float> t0 = Avx.Add(s.V0, s.V7);
-            Vector256<float> t7 = Avx.Subtract(s.V0, s.V7);
-            Vector256<float> t1 = Avx.Add(s.V1, s.V6);
-            Vector256<float> t6 = Avx.Subtract(s.V1, s.V6);
-            Vector256<float> t2 = Avx.Add(s.V2, s.V5);
-            Vector256<float> t5 = Avx.Subtract(s.V2, s.V5);
-            Vector256<float> t3 = Avx.Add(s.V3, s.V4);
-            Vector256<float> t4 = Avx.Subtract(s.V3, s.V4);
-
-            Vector256<float> c0 = Avx.Add(t0, t3);
-            Vector256<float> c1 = Avx.Add(t1, t2);
-
-            // 0 4
-            d.V0 = Avx.Add(c0, c1);
-            d.V4 = Avx.Subtract(c0, c1);
-
-            Vector256<float> c3 = Avx.Subtract(t0, t3);
-            Vector256<float> c2 = Avx.Subtract(t1, t2);
+            block.Transpose();
+            IDCT8x8(ref block, ref temp);
+            temp.Transpose();
+            IDCT8x8(ref temp, ref block);
 
-            // 2 6
-            d.V2 = SimdUtils.HwIntrinsics.MultiplyAdd(Avx.Multiply(c2, C_V_0_5411), c3, C_V_1_3065);
-            d.V6 = SimdUtils.HwIntrinsics.MultiplySubstract(Avx.Multiply(c2, C_V_1_3065), c3, C_V_0_5411);
-
-            c3 = SimdUtils.HwIntrinsics.MultiplyAdd(Avx.Multiply(t4, C_V_1_1758), t7, C_V_0_7856);
-            c0 = SimdUtils.HwIntrinsics.MultiplySubstract(Avx.Multiply(t4, C_V_0_7856), t7, C_V_1_1758);
-
-            c2 = SimdUtils.HwIntrinsics.MultiplyAdd(Avx.Multiply(t5, C_V_1_3870), C_V_0_2758, t6);
-            c1 = SimdUtils.HwIntrinsics.MultiplySubstract(Avx.Multiply(C_V_0_2758, t5), t6, C_V_1_3870);
-
-            // 3 5
-            d.V3 = Avx.Subtract(c0, c2);
-            d.V5 = Avx.Subtract(c3, c1);
-
-            c0 = Avx.Multiply(Avx.Add(c0, c2), C_V_InvSqrt2);
-            c3 = Avx.Multiply(Avx.Add(c3, c1), C_V_InvSqrt2);
-
-            // 1 7
-            d.V1 = Avx.Add(c0, c3);
-            d.V7 = Avx.Subtract(c0, c3);
-#endif
+            // TODO: This can be fused into quantization table step
+            block.MultiplyInPlace(C_0_125);
         }
 
         /// <summary>
-        /// Performs 8x8 matrix Forward Discrete Cosine Transform
+        /// Apply 2D floating point FDCT inplace.
         /// </summary>
-        /// <param name="s">Source</param>
-        /// <param name="d">Destination</param>
-        public static void FDCT8x8(ref Block8x8F s, ref Block8x8F d)
+        /// <param name="block">Input matrix.</param>
+        public static void TransformFDCT(ref Block8x8F block)
         {
 #if SUPPORTS_RUNTIME_INTRINSICS
-            if (Avx.IsSupported)
+            if (Sse.IsSupported)
             {
-                FDCT8x8_Avx(ref s, ref d);
+                ForwardTransformSimd(ref block);
             }
             else
 #endif
             {
-                FDCT8x4_LeftPart(ref s, ref d);
-                FDCT8x4_RightPart(ref s, ref d);
```

Comment:
Maybe we can save `Vector4` implementation for older frameworks via defines before 2.0 and remove it post 2.0? Would look ugly as hell but at least it can be hidden in a single method which would call either SIMD or `Vector4` implementation.

Reply:
See https://github.com/SixLabors/ImageSharp/pull/1761#issuecomment-927270457:

I would consider going `Vector4` for everything non-AVX, since it provides better compatibility (and will provide so), while being actually SSE.

Reply:
For v3 all the platforms we support will support the `Runtime.Intrinsics` API. (_NET Core 3.1+_) I would keep both approaches for now and drop the `Vector4` approach simply for readability. (_We want the same pattern for SSE and AVX, SIMD is hard enough to read as it is_). There wouldn't be any performance loss at that time .

Reply:
I somewhat disagree here. In ideal case, `System.Runtime.Intrinsics` is only superior to `System.Numerics` vectors if the algorithmic optimizations are architecture-specific, which is not the case with this DCT implementation, since it uses simple arithmetic operators only.

The runtime team may decide to implement ARM support for `Vector4` in the future, meaning that our code will get perf benefits on platforms without us changing it. I can't assess how much is this thinking realistic in practice, will try to get some advice later today when US awakes.

Additionally, if really want to avoid regressions on legacy 2.0 targets, DCT is a critical area.

Reply:
Actually, ARM64 support for `System.Numerics` is already a thing, scroll to System.Numerics here:

https://devblogs.microsoft.com/dotnet/arm64-performance-in-net-5/

Reply:
Good point. I'm being a little selfish tbh, I struggle with SIMD so find repeated patterns to be useful. On a side note, I wish we had am ARM test runner!

Reply:
To conclude: `Vector4` for sse and explicit simd for avx, right?

Reply:
Yes.